File: src/swingjs/plaf/JSRadioButtonUI.java
package swingjs.plaf;

import java.util.HashMap;
import java.util.Map;

import swingjs.api.DOMNode;
import jsjava.awt.Dimension;
import jsjavax.swing.AbstractButton;
import jsjavax.swing.ButtonGroup;
import jsjavax.swing.DefaultButtonModel;
import jsjavax.swing.ImageIcon;
import jsjavax.swing.JRadioButton;

public class JSRadioButtonUI extends JSButtonUI {

	private DOMNode label;
	private DOMNode wrapper;
	private static Map<ButtonGroup, String> groupNames;
	

	@Override
	protected DOMNode updateDOMNode() {
		isRadio = true;
		return getButtonObject("radio");
	}

	@Override
	protected String getPropertyPrefix() {
		return "RadioButton.";
	}

	protected DOMNode getButtonObject(String myType) {
		JRadioButton b = (JRadioButton) c;
		boolean isNew = false;
		boolean doAll = false;
		if (domNode == null) {
			doAll = true;
			if (groupNames == null)
				groupNames = new HashMap<ButtonGroup, String>();
			ButtonGroup bg = null;
			String name = id;
			if (b.getModel() instanceof DefaultButtonModel) {
				bg = ((DefaultButtonModel) b.getModel()).getGroup();
				name = groupNames.get(bg);
				isNew = (bg != null && name == null);
				if (name == null)
					name = id;
				if (isNew)
					groupNames.put(bg, name);
			}
			domBtn = enableNode = newDOMObject("input", id, "type", myType, "name",
					name);
			setDataComponent(domBtn);
			iconNode = newDOMObject("span", id + "_icon");
			textNode = newDOMObject("label", id + "l1");
			setEnabled(c.isEnabled());
			label = newDOMObject("label", id + "l2", "htmlFor", id);
			setDataComponent(label);
			setDataComponent(iconNode); // needed for mac safari/chrome
			setDataComponent(textNode); // needed for mac safari/chrome
			// hasOuterDiv will be true unless this is a menu bar menu ?
			if (hasOuterDiv) {
				wrapper = label;
			} else {
				wrapper = createItem("_item", label);
			}
		}
		if (b.isSelected() || isNew)
			DOMNode.setAttr(domBtn, "checked", "true");
		setCssFont(textNode, c.getFont());

		setIconAndText("radio", (ImageIcon) button.getIcon(), button.getIconTextGap(), button.getText());

		// Get the dimensions of the radio button by itself.

		// We have to remove any position:abolute because if that
		// is there, it messes up the width and height calculation.
		DOMNode.setStyles(iconNode, "position", null);
		DOMNode.setStyles(domBtn, "position", null);
		DOMNode.setStyles(textNode, "position", null);
		DOMNode obj;
		Dimension dobj = null;
		if (hasOuterDiv) {
			// We need the width of the text to position the button.
			// But we need to slightly underestimate it so that the
			// width of label + button does not go over the total calculated width
			int wBtn = setHTMLSize1(domBtn, false, false).width - 1;
			int wIcon = Math.max(0, setHTMLSize1(iconNode, false, false).width - 1);
			// Now wrap the two with a div and get its dimensions
			// and then put them back into the wrapper.
			dobj = setHTMLSize1(wrap("div", "", iconNode, domBtn, textNode), false, false);
			// set the offset of the text based on the icon and radio button size
			DOMNode.setStyles(textNode, "left", wBtn + wIcon + "px");
			// add a couple of vertical adjustments

			vCenter(domBtn, -75);
			vCenter(iconNode, -15);
			vCenter(textNode, -50);

			// make everything absolute to pass sizing info to all
			DOMNode.setPositionAbsolute(domBtn, Integer.MIN_VALUE, 0);
			DOMNode.setPositionAbsolute(iconNode, Integer.MIN_VALUE, 0);
			DOMNode.setPositionAbsolute(textNode, Integer.MIN_VALUE, 0);
			DOMNode.setPositionAbsolute(label, Integer.MIN_VALUE, 0);

			// (Re)create label.
		}
		label.appendChild(iconNode);
		label.appendChild(domBtn);
		label.appendChild(textNode);
		if (doAll) {
			// now wrap these in a div
			domNode = obj = (hasOuterDiv ? wrap("div", id + "_0", label) : wrapper);
			if (hasOuterDiv)
				DOMNode.setPositionAbsolute(obj, Integer.MIN_VALUE, 0);
		} else {
			obj = domNode;
		}
		if (hasOuterDiv) {
			DOMNode.setSize(label, dobj.width, dobj.height);
			DOMNode.setSize(obj, dobj.width, dobj.height);
		}
		return obj;
	}

	@Override
	protected Dimension setHTMLSize(DOMNode obj, boolean addCSS) {
		// "absolute" is required for positioning of button, but must not be there
		// for setting the size.
		DOMNode.setStyles(label, "position", null, "width", null, "height", null);
		DOMNode.setStyles(domBtn, "position", null, "width", null, "height", null);
		DOMNode
				.setStyles(textNode, "position", null, "width", null, "height", null);
		Dimension d;
		if (hasOuterDiv) {
			d = setHTMLSize1(obj, addCSS, false);
			DOMNode.setPositionAbsolute(domBtn, Integer.MIN_VALUE, 0);
			DOMNode.setPositionAbsolute(textNode, Integer.MIN_VALUE, 0);
			DOMNode.setPositionAbsolute(label, Integer.MIN_VALUE, 0);
			DOMNode.setStyles(label, "width", d.width + "px", "height", d.height
					+ "px");
		} else {
			d = new Dimension(20,20);
		}
		return d;
	}


	public void handleDOMEvent(Object e) {
		((AbstractButton) c).doClick(0);
	}
	
	@Override
	protected int getDefaultIconTextGap() {
		return 4;
	}


}

File: site/swingjs/j2s/swingjs/plaf/JSRadioButtonUI.js
Clazz.declarePackage ("swingjs.plaf");
Clazz.load (["swingjs.plaf.JSButtonUI"], "swingjs.plaf.JSRadioButtonUI", ["java.util.HashMap", "javax.swing.DefaultButtonModel", "swingjs.api.DOMNode"], function () {
c$ = Clazz.decorateAsClass (function () {
this.wrapper = null;
Clazz.instantialize (this, arguments);
}, swingjs.plaf, "JSRadioButtonUI", swingjs.plaf.JSButtonUI);
Clazz.overrideMethod (c$, "getDOMObject", 
function () {
this.isRadio = true;
return this.getButtonObject ("radio");
});
Clazz.overrideMethod (c$, "getPropertyPrefix", 
function () {
return "RadioButton.";
});
Clazz.overrideMethod (c$, "setHTMLSize", 
function (obj, addCSS) {
swingjs.api.DOMNode.setStyles (this.domBtn, ["position", null]);
swingjs.api.DOMNode.setStyles (this.textNode, ["position", null]);
swingjs.api.DOMNode.setStyles (this.wrapper, ["position", null]);
var d = this.setHTMLSize1 (obj, addCSS, false);
swingjs.api.DOMNode.setPositionAbsolute (this.domBtn);
swingjs.api.DOMNode.setPositionAbsolute (this.textNode);
swingjs.api.DOMNode.setPositionAbsolute (this.wrapper);
return d;
}, "swingjs.api.DOMNode,~B");
Clazz.defineMethod (c$, "getButtonObject", 
function (myType) {
var b = this.c;
var isNew = false;
var doAll = false;
if (this.domNode == null) {
doAll = true;
if (swingjs.plaf.JSRadioButtonUI.groupNames == null) swingjs.plaf.JSRadioButtonUI.groupNames =  new java.util.HashMap ();
var bg = null;
var name = this.id;
isNew = true;
if (Clazz.instanceOf (b.getModel (), javax.swing.DefaultButtonModel)) {
bg = (b.getModel ()).getGroup ();
name = swingjs.plaf.JSRadioButtonUI.groupNames.get (bg);
if (name == null) swingjs.plaf.JSRadioButtonUI.groupNames.put (bg, name = this.id);
 else isNew = false;
}this.domBtn = this.enableNode = this.createDOMObject ("input", this.id, ["type", myType, "name", name]);
this.textNode = this.createDOMObject ("label", this.id + "l", []);
this.wrapper = this.createDOMObject ("label", this.id + "2", ["htmlFor", this.id]);
swingjs.api.DOMNode.add (this.wrapper, this.domBtn);
swingjs.api.DOMNode.add (this.wrapper, this.textNode);
}if (b.isSelected () || isNew) swingjs.api.DOMNode.setAttr (this.domBtn, "checked", "true");
this.setCssFont (swingjs.api.DOMNode.setAttr (this.textNode, "innerHTML", (this.c).getText ()), this.c.getFont ());
var drad = this.setHTMLSize1 (this.domBtn, false, false);
this.setHTMLSize1 (this.textNode, false, false);
var obj = this.wrap ("div", "", [this.domBtn, this.textNode]);
var dobj = this.setHTMLSize1 (obj, true, true);
swingjs.api.DOMNode.add (this.wrapper, this.domBtn);
swingjs.api.DOMNode.add (this.wrapper, this.textNode);
swingjs.api.DOMNode.setStyles (this.textNode, ["left", drad.width + "px"]);
swingjs.plaf.JSComponentUI.vCenter (this.domBtn, -75);
swingjs.plaf.JSComponentUI.vCenter (this.textNode, -50);
swingjs.api.DOMNode.setPositionAbsolute (this.domBtn);
swingjs.api.DOMNode.setPositionAbsolute (this.textNode);
swingjs.api.DOMNode.setPositionAbsolute (this.wrapper);
if (doAll) {
obj = this.wrap ("div", this.id + "_0", [this.wrapper]);
swingjs.api.DOMNode.setPositionAbsolute (obj);
} else {
obj = this.domNode;
}swingjs.api.DOMNode.setSize (this.wrapper, dobj.width, dobj.height);
return swingjs.api.DOMNode.setSize (obj, dobj.width, dobj.height);
}, "~S");
Clazz.defineStatics (c$,
"groupNames", null);
});
